File: OCThread/OCThread/TimerVc.m
//
//  TimerVc.m
//  OCThread
//
//  Created by ZGY on 2018/1/4.
//Copyright © 2018年 macpro. All rights reserved.
//
//  Author:        Airfight
//  My GitHub:     https://github.com/airfight
//  My Blog:       http://airfight.github.io/
//  My Jane book:  http://www.jianshu.com/users/17d6a01e3361
//  Current Time:  2018/1/4  下午12:56
//  GiantForJade:  Efforts to do my best
//  Real developers ship.

#import "TimerVc.h"
#import "GYWeakProxy.h"
// Controllers

// Model

// Views

//#define <#macro#> <#value#>

@interface TimerVc ()

@property (nonatomic,strong) NSTimer *timer;

@end

@implementation TimerVc

- (void)viewDidLoad {
    [super viewDidLoad];
    GYWeakProxy *PROXY = [GYWeakProxy proxyWithTarget:self];
    _timer = [NSTimer scheduledTimerWithTimeInterval:1 target:PROXY selector:@selector(run) userInfo:nil repeats:true];
    [[NSRunLoop currentRunLoop] addTimer:_timer forMode:NSDefaultRunLoopMode];
    
    [self performSelector:@selector(invalide) withObject:nil afterDelay:3];
    
}

- (void)run
{
    NSLog(@"run");
}

- (void)invalide
{
    [_timer invalidate];
    _timer = nil;
}
- (void)dealloc
{

    [_timer invalidate];
    _timer = nil;
}



#pragma mark - Override


#pragma mark - Initial Methods


#pragma mark - Delegate


#pragma mark - Target Methods


#pragma mark - Notification Methods


#pragma mark - KVO Methods


#pragma mark - UITableViewDelegate, UITableViewDataSource


#pragma mark - Privater Methods


#pragma mark - Setter Getter Methods


#pragma mark - Life cycle

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
}

/*
- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}
*/

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SwiftThread/SwiftThread/TimerVc.swift
//
//  TimerVc.swift
//  SwiftThread
//
//  Created by ZGY on 2018/1/4.
//Copyright © 2018年 macpro. All rights reserved.
//
//  Author:        Airfight
//  My GitHub:     https://github.com/airfight
//  My Blog:       http://airfight.github.io/
//  My Jane book:  http://www.jianshu.com/users/17d6a01e3361
//  Current Time:  2018/1/4  下午1:04
//  GiantForJade:  Efforts to do my best
//  Real developers ship.

import UIKit

protocol Pizzeria {
    func makePizza(_ ingredients:[String])
    func makeMargherita()
}

extension Pizzeria {
    func makeMargherita() {
        return makePizza(["1"])
    }
}

struct Lomabards:Pizzeria {
    func makeMargherita() {
        return makePizza(["2"])
    }
    
    func makePizza(_ ingredients: [String]) {
        print(ingredients)
    }
}

class TimerVc: UIViewController {
    
    //MARK: - Attributes
    
    var timer:Timer!
    override func viewDidLoad() {
        super.viewDidLoad()
//        timer = Timer(timeInterval: 1, target: self, selector: #selector(run), userInfo: nil, repeats: true)
//        RunLoop.current.add(timer, forMode: RunLoopMode.defaultRunLoopMode)
//        timer.fire()
        
    }
    
    @objc func run() {
        print("run")
    }
    
    deinit {
//        timer.invalidate()
//        timer = nil
        print("销毁了")
    }
    
    //MARK: - Override
    
    
    //MARK: - Initial Methods
    
    
    //MARK: - Delegate
    
    
    //MARK: - Target Methods
    
    
    //MARK: - Notification Methods
    
    
    //MARK: - KVO Methods
    
    
    //MARK: - UITableViewDelegate, UITableViewDataSource
    
    
    //MARK: - Privater Methods
    
    
    //MARK: - Setter Getter Methods
    
    
    //MARK: - Life cycle
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
    }

    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}
